server/src/main/kotlin/suwayomi/tachidesk/manga/impl/download/DownloadManager.kt

```diff
@@ -75,24 +80,77 @@ object DownloadManager {
-                            ChapterTable.select { (ChapterTable.manga eq mangaId) and (ChapterTable.sourceOrder eq chapterIndex) }
-                                .first()
+    @Serializable
+    data class EnqueueInput(
+        val mangaId: Int,
+        val chapterIndex: Int
+    )
+
+    fun enqueueMultiple(inputs: List<EnqueueInput>) {
+        val mangas = transaction {
+            val mangaIds = inputs.map { it.mangaId }.distinct()
+            MangaTable.select { MangaTable.id inList mangaIds }
+                .map { MangaTable.toDataClass(it) }
+        }
+
+        // This list will have unwanted chapters from other manga but there is no simple way
+        // to select only wanted ones. It will be mapped to input later so it should not be problem
+        val chapters = transaction {
+            val chapterIndexes = inputs.map { it.chapterIndex }.distinct()
+            ChapterTable.select { ChapterTable.sourceOrder inList chapterIndexes }.toList()
+        }
+
+        val mappedInputs = transaction {
+            inputs.map {
+                Pair(
+                    mangas.first { manga -> manga.id == it.mangaId },
+                    ChapterTable.toDataClass(
+                        chapters.first { chapter ->
+                            inputs.find { input ->
+                                it.chapterIndex == chapter[ChapterTable.sourceOrder] &&
+                                    input.mangaId == chapter[ChapterTable.manga].value
+                            } != null
                         }
-                    ),
-                    manga = getManga(mangaId)
+                    )
                 )
-            )
+            }
+        }
+
+        addMultipleToQueue(mappedInputs)
+    }
+
+    /**
+     * Tries to add multiple inputs to queue
+     * If any of inputs was actually added to queue, starts the queue
+     */
+    private fun addMultipleToQueue(inputs: List<Pair<MangaDataClass, ChapterDataClass>>) {
+        val addedChapters = inputs.map { addToQueue(it.first, it.second) }
+        val anyAdded = addedChapters.any { it != null }
+
+        if (anyAdded) {
             start()
+            notifyAllClients()
         }
-        notifyAllClients()
+    }
+
+    /**
+     * Tries to add chapter to queue.
+     * If chapter is added, returns the created DownloadChapter, otherwise returns null
+     */
+    private fun addToQueue(manga: MangaDataClass, chapter: ChapterDataClass): DownloadChapter? {
+        if (downloadQueue.none { it.mangaId == manga.id && it.chapterIndex == chapter.index }) {
+            val dc = DownloadChapter(
+                chapter.index,
+                manga.id,
+                chapter,
+                manga
+            )
+            downloadQueue.add(dc)
+            logger.debug("Added to download queue: ${manga.title} | ${chapter.index}")
+            return dc
+        }
+        logger.debug("Chapter already present in queue: ${manga.title} | ${chapter.index}")
```

Comment:
This helps with performance when debug logging is not enabled
```suggestion
        logger.debug { "Chapter already present in queue: ${manga.title} | ${chapter.index}" }
```